packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1445,10 +1445,22 @@ export class KeyringController extends BaseController<
    * @returns Promise resolving when the operation completes.
    */
   async submitPassword(password: string): Promise<void> {
-    return this.#withRollback(async () => {
+    await this.#withRollback(async () => {
       this.#keyrings = await this.#unlockKeyrings(password);
       this.#setUnlocked();
     });
+
+    try {
+      // If there are stronger encryption params available, we
+      // can attempt to upgrade the vault encryption params.
+      await this.#withRollback(async () =>
```

Comment:
Using this wrapper ensures that:
- The upgrade is mutually exclusive
- Everything gets rolled back to the exact point right before its execution, even if something else is executed between the above login promise and this other promise